### PR TITLE
Publish events to RabbitMQ

### PR DESCRIPTION
**Pull request by oyvindhagberg**
Events are sent to MQ when:
- Hosts are added or removed
- Roles or atoms are created or deleted
- Atoms are added to or removed from roles

### CODE REVIEW

File: mreg/utils.py [hunk 1]
```diff
@@ -130,3 +140,59 @@ def normalize_mac(mac: str) -> str:
     mac = re.sub('[.:-]', '', mac).lower()
     return ":".join(["%s" % (mac[i:i+2]) for i in range(0, 12, 2)])
 
+
+mq_channel = None
+mq_id : int = 1
+pid_and_time = str(os.getpid()) + " " + str(time.time()) + " "
+
+def send_event_to_mq(obj, routing_key):
```
[the remark this answers is not in the document]
Comment (oyvindkolbu) [reply mreg/utils.py:148]: Helt enig med @mbakke om egen `mq.py`, samt gjerne ting egen klasse.

File: mreg/utils.py [hunk 2]
```diff
@@ -130,3 +140,59 @@ def normalize_mac(mac: str) -> str:
     mac = re.sub('[.:-]', '', mac).lower()
     return ":".join(["%s" % (mac[i:i+2]) for i in range(0, 12, 2)])
 
+
+mq_channel = None
+mq_id : int = 1
+pid_and_time = str(os.getpid()) + " " + str(time.time()) + " "
```
Comment (oyvindkolbu) [inline mreg/utils.py:146]: Hvorfor statisk variabel med et fast tidspunkt?

File: mreg/utils.py [hunk 3]
```diff
@@ -130,3 +140,59 @@ def normalize_mac(mac: str) -> str:
...
+    config = getattr(settings, 'MQ_CONFIG', None)
+    if config is None:
+        return
+
+    global mq_channel
+    global mq_id
+    global pid_and_time
+
+    # Add an id property to the event
+    obj['id'] = hashlib.md5((pid_and_time + str(mq_id)).encode('utf-8')).hexdigest()
```
Comment (oyvindkolbu) [inline mreg/utils.py:158]: Med mindre dette må være en md5-digest, så bør ikke noe mye kode bruke md5, samme hvor trivielt.

File: mreg/utils.py (hunk 2 above)
Comment (oyvindhagberg) [reply mreg/utils.py:146]: Så koden slipper å kalle opp os.getpid() og time.time() hver gang en event oppstår.
Og grunnen til at pid og tidspunktet er en del av event-id'ene er at hvis mreg restartes så vil vi unngå at event-id'er kan gjenta seg. Samtidig vil vi unngå å ha et løpenummer som vi må lagre noe sted når mreg avsluttes.

Comment (oyvindhagberg) [reply mreg/utils.py]: Endret i 76759f2

File: mreg/utils.py (hunk 3 above)
Comment (oyvindhagberg) [reply mreg/utils.py]: Fjernet md5 i 76759f2

File: mreg/mqsender.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+class MQSender(object):
+
+	def __new__(cls):
+		if not hasattr(cls, 'instance'):
+			cls.instance = super(MQSender, cls).__new__(cls)
+		return cls.instance
+
+	def __init__(self):
+		self.mq_channel = None
+		self.mq_id : int = int(time.time_ns()/1000000)
```
Comment (oyvindkolbu) [inline mreg/mqsender.py:24]: No need to force the int conversion, use instead:
`time.time_ns()//1_000_000`

**Review by oyvindkolbu — APPROVED** [review]
Looks better now. Though tests would be nice, that would require a fair bit of mocking.

**Review by oyvindkolbu — APPROVED** [review]
:shipit:

**Comment by ponas** [comment]
Some downsides to this implementation:
* if the message broker is having performance issues, it will latency to all model changes / API requests
  * best case: the connection is retried ten times
  * worst case: the connection hangs, waiting for a timeout
* if the message broker is down, messages are silently discarded

Suggestion:
* add an intermediate `Notification` model for serialized message data
* change the signals to create `Notification` objects
* add a long-running process (via a management command) that:
  * selects the N oldest notifications
  * publishes them to the message broker
  * deletes the notification if successful
  * sleeps, repeats

With this, an auto-incrementing primary key could be used as the message ID.

**Pull request merged**